source/common/http/filter/lua/wrappers.cc

```diff
@@ -33,7 +33,22 @@ int HeaderMapWrapper::luaAdd(lua_State* state) {
 
   const char* key = luaL_checkstring(state, 2);
   const char* value = luaL_checkstring(state, 3);
-  headers_.addCopy(LowerCaseString(key), value);
+  const LowerCaseString lower_case_key(key);
+
+  // We handle inline headers differently from non-inline headers. Since we only offer the add()
+  // API, users generally expect the header to either be added or overwritten. This is not how
+  // the inline headers currently work where the first version is taken to be the authoritative
+  // version and the rest dropped during external message parse. Internally, code operates on
+  // O(1) headers directly so we don't have this problem in the rest of Envoy. In the Lua case, we
+  // handle this by checking if the header in question is an O(1) header. If it is, we just set the
+  // value directly. Otherwise we add it per normal.
```

Comment:
So the new behavior is if you "add" a header which is defined in ALL_INLINE_HEADERS it will replace, and if you "add" a header not defined in ALL_INLINE_HEADERS it will (functionally) append?    Are inline headers supposed to be headers which are not allowed to be comma-delimited?  At a glance I see at least one ForwardedFor I believe often has multiple values and having lua sometimes replace and sometimes append strikes me as more inconsistent and confusing, and merits some docs.

I think it'd be more clear if we had an add() and an append() and users explicitly called the one they wanted.

Reply:
Yeah, I agree this is a pretty confusing situation. There is an open issue on the internal lack of append functionality for inline headers: https://github.com/envoyproxy/envoy/issues/2004.

IIRC `x-forwarded-for` is the only legitimate case where multiple values should probably be supported and switched to appending with `,`.

I'm torn on what to do here. This is not a problem in C++ code as people access the inline headers directly so there is no confusion. This is not possible from Lua. My intent here was to try to paper over the differences in the common case. So I guess options are:
1) I add better docs in data-plane-api about this, and keep this change.
2) Do not do this change, and more clearly doc current behavior. Then potentially add `replace` API which people can use to do the remove/add combo more easily.

WDYT?

Reply:
At first pass I was neutral but I think if we stick with 1) anyone adding a new inline header would then have their "performance optimization" result in an unexpected and possibly detrimental change for lua users, so I think I lean towards 2) as I think it may save us some pain down the line.   

We could also go with 3) where add "does the right thing", appending to append-able headers, replacing single value headers, and this fixes one bug and introduces a second (x-forward-for) but I think I still prefer explicit add/replace semantics as more flexible.

Reply:
Alright I will do (2).